rust/lance-linalg/src/simd/f16.c

```diff
@@ -55,3 +56,26 @@ float l2_f16(const _Float16 *x, const _Float16 *y, uint32_t dimension) {
   }
   return (float) sum;
 }
+
+/// @brief Fast cosine function, that assumes that the norm of the first vector is already known.
+/// @param x A f16 vector
+/// @param x_norm Norm of the first vector
+/// @param y A f16 vector
+/// @param dimension The dimension of the vectors
+/// @return The cosine distance of the two vectors.
+float cosine_fast_f16(const _Float16 *x, const float x_norm, const _Float16 *y, uint32_t dimension) {
+  _Float16 xy = 0.0;
+  _Float16 y_sq = 0.0;
```

Comment:
If we make these both `float`, we still get SIMD instruction and much shorter generated assembly. This would give us much more precise values, which might be worth it. Could you try that and benchmark against the existing implementation?

Reply:
I'm seeing [float version](https://godbolt.org/z/5dEzfc9vK) longer (320 lines) than the [f16 version](https://godbolt.org/z/Kb57P7esT) (286 lines), but the percission might well be worth it. Will benchmark.

Reply:
Okay nevermind, I must have knocked the O3 off accidentally.